### PR TITLE
[tool] Fix flutter.js regression with hot-reload on promise-based init.

### LINKED ISSUE #109093: Hot Restart broken on Flutter Web  with `--web-renderer html`
So I Updated my Flutter today. 

I can run my Flutter Web with 

> flutter run -d chrome --web-renderer html

without a problem. But when I want to hot restart it I just get a white screen. 

These are all of the versions of flutter, dart etc. that Im using

> Flutter 3.1.0-0.0.pre.2127 • channel master •
> https://github.com/flutter/flutter.git
> Framework • revision 792a5829e9 (2 hours ago) • 2022-08-06 13:20:08
> -0400
> Engine • revision a195b6b6f3
> Tools • Dart 2.19.0 (build 2.19.0-72.0.dev) • DevTools 2.16.0

I even created a whole new flutter web project with:

> flutter create testigo
> cd testigo
> flutter run -d chrome --web-renderer html

 and the problem persisted. So I saw the default starting screen of any new flutter project but it was not responsive. I couldn't click the increment button after hot restart

I ran flutter doctor and everything seems good as well:

> 
> [✓] Flutter (Channel master, 3.1.0-0.0.pre.2127, on macOS 12.5 21G72 darwin-x64,
>     locale en-GB)
> [✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
> [!] Xcode - develop for iOS and macOS
>     ✗ Xcode installation is incomplete; a full installation is necessary for iOS
>       development.
>       Download at: https://developer.apple.com/xcode/download/
>       Or install Xcode via the App Store.
>       Once installed, run:
>         sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
>         sudo xcodebuild -runFirstLaunch
> [✓] Chrome - develop for the web
> [✓] Android Studio (version 2021.1)
> [✓] VS Code (version 1.70.0)
> [✓] Connected device (2 available)
> [✓] HTTP Host Availability
>

### PR DESCRIPTION
**Pull request by ditman**
This PR changes `flutter.js` to fix hot-reload on promise-based initializations. See details [here](https://github.com/flutter/flutter/issues/109093#issuecomment-1233611777).

This PR also extends the current testing infrastructure so now we can test any app bootstrap (index.html) and ensure that hot reload keeps working.

Fixes https://github.com/flutter/flutter/issues/109093

---
_The first test run had an intentional error, to verify that CI caught the problem, and [it did!](https://github.com/flutter/flutter/pull/110805#issuecomment-1234814177)_

---

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart [hunk 1]
```diff
@@ -205,6 +205,8 @@ _flutter.loader = null;
         this._didCreateEngineInitializerResolve(engineInitializer);
         // Remove the resolver after the first time, so Flutter Web can hot restart.
         this._didCreateEngineInitializerResolve = null;
+        // Make the engine revert to "auto" initialization on hot restart.
+        // delete _flutter.loader.didCreateEngineInitializer;
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart:209]: LGTM, once this line is uncommented :)

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart (hunk 1 above)
Comment (ditman) [reply packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart:209]: Yes, I want to see this fail, then pass in CI.

I've run the tests locally, but I want to make sure that I'm not breaking anything weird.

**Comment by ditman** [comment]
OK, it seems that the `web_tool_tests` are failing as expected:

```
...

03:02 +23 ~3 -1: test/web.shard/hot_reload_web_test.dart: Hot reload (index.html: flutter.js (promises)): newly added code executes during hot restart [E]                                             
  TimeoutException after 0:00:15.000000: Future not completed

...

06:41 +27 ~3 -2: test/web.shard/hot_reload_web_test.dart: Hot reload (index.html: flutter.js (promises, short)): newly added code executes during hot restart [E]                                      
  TimeoutException after 0:00:15.000000: Future not completed
  
To run this test again: /b/s/w/ir/x/w/flutter/bin/cache/dart-sdk/bin/dart test test/web.shard/hot_reload_web_test.dart -p vm --plain-name 'Hot reload (index.html: flutter.js (promises, short)): newly added code executes during hot restart'

...

20:25 +51 ~7 -2: Some tests failed.                                                                                                                                                                    

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
╔═╡ERROR╞═══════════════════════════════════════════════════════════════════════
║ Command exited with exit code 1 but expected: zero exit code.
║ Command: ../../bin/cache/dart-sdk/bin/dart run test --test-randomize-ordering-seed=20220901 -j1 --no-color test/web.shard
║ Relative working directory: packages/flutter_tools
╚═══════════════════════════════════════════════════════════════════════════════
▌14:22:01▐ Test failed. 

```

[Build log](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8804191277042270977/+/u/run_test.dart_for_web_tool_tests_shard_and_subshard_web/test_stdout)

I'll wait for all steps to complete (this should fail regardless of the OS), and push the fix.

**Review by christopherfujino — APPROVED** [review]
LGTM, looks like your test (correctly) failed on CI

**Comment by ditman** [comment]
All right, this fails as expected across all platforms:

* Linux: https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20web_tool_tests/19243/overview
* Mac: https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20web_tool_tests/18911/overview
* Windows: https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20web_tool_tests/15661/overview

Pushing the fix.

**Comment by ditman** [comment]
Succeeding now:

Linux: 🟡 https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20web_tool_tests/19246/overview
Mac: 🟢 https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20web_tool_tests/18914/overview
Windows: 🟡 https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20web_tool_tests/15664/overview

**Comment by ditman** [comment]
Adding `autosubmit` to land this.

**Pull request merged**